firedrake/adjoint_utils/blocks/function.py

```diff
@@ -139,7 +137,7 @@ def evaluate_tlm_component(self, inputs, tlm_inputs, block_variable, idx,
                 dudmi.assign(ufl.algorithms.expand_derivatives(
                     ufl.derivative(expr, dep.saved_output,
                                    dep.tlm_value)))
-                dudm.vector().axpy(1.0, dudmi.vector())
+                dudm.dat += 1.0 * dudmi.dat
```

Comment:
I've never seen direct operations on a `PyOP2.dat` like this before. @connorjward is it the right thing to be doing?

Reply:
This is exactly what `Vector.axpy` used to do. We are just lifting the code to not rely on `Vector`.